File: src/main/java/eu/mshade/enderframe/Watchable.java
package eu.mshade.enderframe;

import java.util.Collection;
import java.util.function.Consumer;

public interface Watchable {

    void addWatcher(Agent agent);

    void removeWatcher(Agent agent);

    boolean isWatching(Agent agent);

    Collection<Agent> getWatchers();

    default void notify(Consumer<Agent> agent){
        for (Agent a : getWatchers()) agent.accept(a);
    }

    default <T extends Agent> void notify(Class<T> agentClass, Consumer<T> agent){
        for (Agent a : getWatchers()) {
            if (agentClass.isInstance(a)) agent.accept((T) a);
        }
    }

}

File: src/main/java/eu/mshade/enderframe/Watchable.kt
package eu.mshade.enderframe

import java.util.function.Consumer

interface Watchable {

    fun addWatcher(agent: Agent?)

    fun removeWatcher(agent: Agent?)

    fun isWatching(agent: Agent?): Boolean

    fun getWatchers(): Collection<Agent>

    fun notify(agent: Consumer<Agent?>?)

    fun <T : Agent?> notify(type: Class<T>, consumer: Consumer<T>) {
        notify { agent: Agent? -> if (type.isInstance(agent)) consumer.accept(type.cast(agent)) }
    }
}